### ISSUE TITLE
About DPE modules

### ISSUE THREAD

**Issue by Xiaobin-Rong** [TungyuYoung/FSPEN#1]
Thanks for your codes!
From my understanding, the **Path extension** method proposed in the original paper is different from **Grouped RNN**. The former method segments the feature across the frequency dimension, whereas the latter divides it along the hidden channel dimension. I'm concerned that in your `dpe.py`, the implementation aligns more with the Grouped RNN approach.

**Comment by TungyuYoung** [comment]
Thanks for pointing that out! I will fix it ASAP!

**Comment by TungyuYoung** [comment]
> Thanks for your codes! From my understanding, the **Path extension** method proposed in the original paper is different from **Grouped RNN**. The former method segments the feature across the frequency dimension, whereas the latter divides it along the hidden channel dimension. I'm concerned that in your `dpe.py`, the implementation aligns more with the Grouped RNN approach.

Dear Rong,
I have reviewd my codes and tried fix the issue. However, I found that there were some parts I do not understand.

As described in the paper, the shape of features output from **Feature Merge** should be [B, T, F_c, C // 2], where F_c denotes the F (257) dimension after encoded should be 32, and the C is 32. Then, in the **Inter GRU**, the authors utilise GRU(16, 16) * 8 the model the feature.  I don't quite understand why these parameters are set as. Perhaps I have misunderstand something. Based on my own codes, dividing the features into eight groups does not seem to match the parameter settings for GRUs as described.

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung 
From my understanding, it splits the feature along the frequency dimension into 8 chunks, and processes each chunk with an independent GRU. Each GRU receives a chunk with a shape of (B*F_c//8, C//2, T). 
I tried implementing the module based on your code.
```
class DPE(nn.Module):
    def __init__(self, input_size, width, hidden_size, **kwargs):
        super().__init__()
        self.input_size = input_size
        self.width = width
        self.hidden_size = hidden_size

        self.intra_rnn = nn.GRU(input_size, hidden_size, num_layers=1, batch_first=True, bidirectional=True)
        self.intra_fc = nn.Linear(32, 16)
        self.intra_ln = nn.LayerNorm((width, hidden_size), eps=1e-8)

        self.inter_rnn = PathExtensionGRU(input_size=input_size, hidden_size=hidden_size, bidirectional=False)

    def forward(self, x):
        """x: (B,C,T,F)"""
        """Intra RNN"""
        x = x.permute(0, 2, 3, 1).contiguous()
        B, T, F, C = x.shape
        intra_x = torch.reshape(x, shape=(B * T, F, C))
        intra_x = self.intra_rnn(intra_x)[0]
        intra_x = self.intra_fc(intra_x)
        intra_x = torch.reshape(intra_x, shape=(B, T, F, C))
        intra_x = self.intra_ln(intra_x)
        intra_out = torch.add(x, intra_x)

        """Inter RNN"""
        x = intra_out.permute(0, 3, 1, 2)  # (B,C,T,F)
        inter_x = self.inter_rnn(x)        # (B,C,T,F)
        dpe_out = torch.add(inter_x, x)  # (B,C,T,F)

        return dpe_out


class PathExtensionGRU(nn.Module):
    def __init__(self, input_size, hidden_size, num_layers=1, batch_first=True, bidirectional=False, chunks=8):
        super().__init__()
        self.chunks = chunks
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        self.bidirectional = bidirectional
        self.rnns = nn.ModuleList([])
        self.fcs = nn.ModuleList([])
        for i in range(chunks):
            self.rnns.append(nn.GRU(input_size, hidden_size, num_layers, batch_first=batch_first, bidirectional=bidirectional))
            self.fcs.append(nn.Linear(hidden_size, hidden_size))


    def forward(self, x):
        """x: (B,C,T,F)"""
        B,C,T,F = x.shape
        x_chunks = torch.chunk(x, self.chunks, dim=3)  # 8 x (B,C,T,F//8)
        y_chunks = []
        for (i, x_chunk) in enumerate(x_chunks):
            x_chunk = x_chunk.permute(0,3,2,1).reshape(-1,T,C)  # (BF//8,T,C)
            y_chunk = self.rnns[i](x_chunk)[0]  # (BF//8,T,C)
            y_chunk = self.fcs[i](y_chunk)      # (BF//8,T,C)
            y_chunk = y_chunk.reshape(B,-1,T,C).permute(0,3,2,1)  # (B,C,T,F//8)
            y_chunks.append(y_chunk)
        y = torch.cat(y_chunks, dim=3)
        return y


if __name__ == "__main__":
    pegru = PathExtensionGRU(16, 16)
    x = torch.randn(1, 16, 63, 32)
    y = pegru(x)
    print(y.shape)
    
    
    dpe = DPE(16, 32, 16)
    y = dpe(x)
    print(y.shape)
```

**Comment by TungyuYoung** [comment]
@Xiaobin-Rong 
Dear Rong,
Thanks for your suggestions!

I reviewed the code you suggested, and I believe your code is more reasonable! I have merged your code into my repository. Additionally, I revisited the paper and removed the **Layer Normalization** from the DPE module. Currentl, the parameters is 78.93 K, and MACs is 83.62 M. There may still be some overlooked areas, but I believe this should have successfully reproduced FSPEN. I will publish the training and testing results later.

Thank you once again for your contributions. Please feel free to comment if you have any questions.

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung 
Thank you for your kind words. However, I am concerned about whether removing the **Layer Normalization** is suitable, cause it's indeed used in the original paper. And, according to my experience, Layer Normalization is important for the performance of the dual-path RNN.

**Comment by TungyuYoung** [comment]
@Xiaobin-Rong 
Yes, you are right. I will re-add Layer Normalization. 

What's more, regarding the mismatch in the parameters and MACs of the model I reproduced, I think the reason for this is that I used Conv2d while the original paper was Conv1d. However, from my understanding, employing Conv1d mentioned in the original paper involves some complex shape transformations and calculations. As a result, I use Conv2d directly. I'm not sure if it is correct. If you have any suggestions, please feel free to provide.

I will complete the relevant experiments as soon as possible and update the results.

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung 
I believe that using Conv2d with a kernel size of (1, x) is equivalent to using Conv1d with a kernel size of x; therefore, it may not be the factor causing the mismatch in complexity count.

I found that the SubbandSplit module in your `feature.py` may not align with the original paper. Notice that *"Overlap and padding are used so that the sub-bands within their respective groups have the same number of frequency bins."* in the original paper, I modified it in this manner to allocate sub-bands in each group:
```
from einops import rearrange

class SubBandSplit(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, specs):
        """
        Split the input specs to 5 groups through the frequency axis.
        The frequency bins in each group is {(1~16), (17~32), (33~64), (65~128), (129~257)}.
        The number of sub-bands in groups is {8, 6, 6, 6, 6}
        :param specs: [batch_size, 2, T, F]
        :return: [batch_size, 2, T, f] * 5
        """
        groups_0 = rearrange(specs[..., 0: 16], 'b c t (n f) -> b (c n) t f', n=8)                        # 16 bands -> 8 subbands
        groups_1 = rearrange(pad(specs[..., 16: 32], [1,1]), 'b c t (n f) -> b (c n) t f', n=6)     # 18 bands -> 6 subbands
        groups_2 = rearrange(pad(specs[..., 32: 64], [2,2]), 'b c t (n f) -> b (c n) t f', n=6)     # 32 bands -> 6 subbands
        groups_3 = rearrange(pad(specs[..., 64: 128], [1,1]), 'b c t (n f) -> b (c n) t f', n=6)   # 66 bands -> 6 subbands
        groups_4 = rearrange(pad(specs[..., 128: 257], [1,2]), 'b c t (n f) -> b (c n) t f', n=6) # 132 bands -> 6 subbands

        x = [groups_0, groups_1, groups_2, groups_3, groups_4]
        return x
```
However, the implementation will lead to a small number of frequency bins, for example, 2 in the sub-bands in the 1st group, which will conflict with the large kernel size and stride settings in SubandEncoder. I don't know how can a convolution work for such a narrow sub-band feature. It's really weird. Please feel free to let me know if my understanding is wrong.

**Comment by TungyuYoung** [comment]
@Xiaobin-Rong 
From my understanding, in my **SubBandEncoder** Module, according to the parameter settings of the sub-band encoder provided in the article, the subdivision and encoding of the sub-bands have been essentially completed. For example, the first group of sub-bands is [0:16], with a convolutional kernel size of (1, 4), a stride of (1, 2), and padding of (0, 1). This effectively implements the subdivision and convolutional encoding within this group with a width of 4 and an overlap of 2. The final result is [B, 32, T, 8], where 8 indicates that there are 8 sub-bands in this group.

What's more, in my **FullBandEncoder** and **FullBandDecoder**, the parameters setting is not same as the paper suggested. For example, in the **second** FullBandEncoder layer, I set the stride of (1, 2) while the paper of (1, 4). If follow the paper's setting, I found that the being processed by the FullBandEncoder, the shape of outputs would be [B, C, T, F_c], where F_c the 16. Then, how can it possible to stack with the sub-band feature [B, C, T, 32] in the Feature Merge porcessing???

I still believe that this paper is not a rigorous one and has little practical value. As you can see, even the spelling of words in the title of this paper is incorrect, which is a very childish mistake! "FSPEN: AN ULTRA-LIGHTWEIGHT NETWORK FOR REAL TIME SPEECH **ENAHNCMENT**", ENAHNCMENT??? If I were the reviwer, perhaps I would provide a strongly reject.

To be more serious, I could even say that this paper involves academic misconduct. I might not have time to follow up on this paper because I have many new projects to pursue. 

Nerver mind. I think we can still keep in touch. I think you are very accomplished in the field of speech enhancement, and I hope we can exchange knowledge and learn from each other.

**Comment by Xiaobin-Rong** [comment]
@TungyuYoung
Yes, you're right. Utilizing a convolutional layer with a well-designed kernel size, stride, and padding to generate a specific number of subbands in the output features is convincing. To some extent, this subband generation approach is more closely aligned with traditional subband filter bank methods, such as the Mel filter bank method. Interestingly, in Mel filter banks, higher frequencies correspond to wider intervals with fewer filters. Here, the higher-frequency groups are associated with larger strides and kernel sizes. This idea is enlightening, and thank you for helping me out of my previous misunderstanding.

As you pointed out, there are mistakes in the parameter settings in the original paper, and I agree with you. Moreover, all the repositories for this model on GitHub emphasize this point. By the way, I hadn't even noticed the spelling mistakes in the title; I admit that it's a really serious issue.

I believe the current implementation has made the utmost effort to remain faithful to the original text. Great job! Thank you for your code and detailed explanations. Looking forward to your test results!
